pageserver/src/relish_storage.rs

```diff
@@ -0,0 +1,103 @@
+//! Abstractions for the page server to store its relish layer data in the external storage.
+//!
+//! Main purpose of this module subtree is to provide a set of abstractions to manage the storage state
+//! in a way, optimal for page server.
+//!
+//! The abstractions hide multiple custom external storage API implementations,
+//! such as AWS S3, local filesystem, etc., located in the submodules.
+
+mod local_fs;
+mod rust_s3;
+/// A queue and the background machinery behind it to upload
+/// local page server layer files to external storage.
+pub mod storage_uploader;
+
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+/// Storage (potentially remote) API to manage its state.
+#[async_trait::async_trait]
+pub trait RelishStorage: Send + Sync {
+    async fn list_relishes(&self) -> anyhow::Result<Vec<Arc<dyn RelishStoragePath>>>;
+
+    async fn download_relish(&self, from: &dyn RelishStoragePath, to: &Path) -> anyhow::Result<()>;
```

Comment:
Not, so far, it's assumed that we've properly configured the pageserver and it's alone watching the directory contents.
So, there would be no concurrent uploads.

The rest of the detection would be done during downloading impl, since we have to not only download the file list, but presumably also parse it and do whatever other comparisons and manipulations that are quite big to squeeze in here IMO.
But good point in general, I've forgotten to add a note on that when created the PR.